<a href="https://colab.research.google.com/github/SonakshiA/Breast-Cancer-using-SVM/blob/main/SVM_Kernel_Comparisons.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install opendatasets

In [2]:
import opendatasets as od

In [3]:
dataset = 'https://www.kaggle.com/datasets/yasserh/breast-cancer-dataset'
od.download(dataset)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: sonakshiarora02
Your Kaggle Key: ··········


100%|██████████| 48.6k/48.6k [00:00<00:00, 15.5MB/s]

In [4]:
import numpy as np
import pandas as pd

In [12]:
breast_cancer_data = pd.read_csv('breast-cancer.csv')
breast_cancer_data.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [7]:
breast_cancer_data['diagnosis'].unique()

array(['M', 'B'], dtype=object)

Separating the labels and datasets

In [14]:
Y = breast_cancer_data['diagnosis']
print(Y)
X = breast_cancer_data.drop(columns='diagnosis')

0      M
1      M
2      M
3      M
4      M
      ..
564    M
565    M
566    M
567    M
568    B
Name: diagnosis, Length: 569, dtype: object


In [15]:
print(X)

           id  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302        17.99         10.38          122.80     1001.0   
1      842517        20.57         17.77          132.90     1326.0   
2    84300903        19.69         21.25          130.00     1203.0   
3    84348301        11.42         20.38           77.58      386.1   
4    84358402        20.29         14.34          135.10     1297.0   
..        ...          ...           ...             ...        ...   
564    926424        21.56         22.39          142.00     1479.0   
565    926682        20.13         28.25          131.20     1261.0   
566    926954        16.60         28.08          108.30      858.1   
567    927241        20.60         29.33          140.10     1265.0   
568     92751         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              

Splitting the data into Training and Testing Data

In [18]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score

In [19]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.3,random_state=100)

Standardising the data to fit within a given range

In [20]:
sc = StandardScaler()
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

Using a **Linear** kernel

In [21]:
classifier = svm.SVC(kernel = 'linear')
classifier.fit(X_train_std,Y_train)

SVC(kernel='linear')

In [22]:
Y_predict = classifier.predict(X_test_std)

In [23]:
print("Testing accuracy: ", accuracy_score(Y_test,Y_predict))

Testing accuracy:  0.9473684210526315


In [25]:
pred_values = pd.DataFrame({'Actual value' : Y_test , 'Predicted value': Y_predict})
pred_values.head()

Actual value Predicted value
400            M               M
225            B               M
321            M               M
173            B               B
506            B               B

In [27]:
from sklearn import metrics
cnf_matrix = metrics.confusion_matrix(Y_test,Y_predict)
print(cnf_matrix)

[[100   2]
 [  7  62]]


Using a **Polynomial** kernel

In [29]:
classifier2 = svm.SVC(kernel = 'poly')
classifier2.fit(X_train_std,Y_train)


SVC(kernel='poly')

In [33]:
Y_predict = classifier2.predict(X_test_std)

In [34]:
print("Testing Accuracy: ", accuracy_score(Y_test,Y_predict))

Testing Accuracy:  0.8947368421052632


In [35]:
cnf_matrix = metrics.confusion_matrix(Y_test,Y_predict)
print(cnf_matrix)

[[102   0]
 [ 18  51]]


Using a **Radial-Basis Function** kernel

In [36]:
classifier3 = svm.SVC(kernel = 'rbf')
classifier3.fit(X_train_std,Y_train)
Y_predict = classifier3.predict(X_test_std)
print("Testing Accuracy: ", accuracy_score(Y_test,Y_predict))

Testing Accuracy:  0.9766081871345029


Using a **Sigmoid** kernel

In [37]:
classifier4 = svm.SVC(kernel = 'sigmoid')
classifier4.fit(X_train_std,Y_train)
Y_predict = classifier4.predict(X_test_std)
print("Testing Accuracy: ", accuracy_score(Y_test,Y_predict))

Testing Accuracy:  0.9298245614035088


In [38]:
cnf_matrix = metrics.confusion_matrix(Y_test,Y_predict)
print(cnf_matrix)

[[100   2]
 [ 10  59]]


**Observations**


1.   Linear Kernel : 0.9473684210526315
2.   Polynmoial Kernel : 0.8947368421052632
3.   RBF Kernel : 0.9766081871345029
4.   Sigmoid Kernel: 0.9298245614035088

We get the highest accuracy with RBF and lowest with Polynomial